xwiki-platform-core/xwiki-platform-repository/xwiki-platform-repository-server-api/src/main/java/org/xwiki/repository/internal/RepositoryManager.java

```diff
@@ -1102,4 +1198,38 @@ private boolean updateCollection(BaseObject extensionObject, String fieldName, C
 
         return needSave;
     }
+
+    public BaseObject getExtensionVersionObject(XWikiDocument extensionDocument, String version) {
+        return getExtensionVersionObject(extensionDocument, version, true);
+    }
+
+    public BaseObject getExtensionVersionObject(XWikiDocument extensionDocument, String version, boolean allowProxying)
+    {
+        if (version == null) {
+            List<BaseObject> objects =
+                    extensionDocument.getXObjects(XWikiRepositoryModel.EXTENSIONVERSION_CLASSREFERENCE);
+
+            if (objects == null || objects.isEmpty()) {
+                return null;
+            } else {
+                return objects.get(objects.size() - 1);
+            }
+        }
+
+        BaseObject extensionVersionObject =
+                extensionDocument.getObject(XWikiRepositoryModel.EXTENSIONVERSION_CLASSNAME, "version", version, false);
+
+        if( extensionVersionObject == null && allowProxying && isVersionProxyingEnabled(extensionDocument) ){
+             //no ExtensionVersionClass object so we need to create such object temporarily
```

Comment:
The idea that I came up with is to - (reusing existing code) add version xObject to document, save document (like it's done so far), then delete this object and save document again. In this way we'll prevent such cases that - this version object is misused whilst manual edit of extension document, or some cache'ing problems. Of course, from the other side - the time for REST access of historic version of extension is getting longer - but I strongly hope it's acceptable. Proxying will regard only previous versions and typically less important extensions - so a bit of delay in such cases is sth we can deal with - I guess.
Let's call it temporary solution - when I get more intuition of how this document, xobject and caching things work, I will rethink it, ok?